File: backend/static/challenge-manager.js
/**
 * Challenge Manager - Frontend
 * 
 * Single source of truth for challenge state with multi-user support.
 * Handles challenge lifecycle, pause-aware timing, and WebSocket synchronization.
 * 
 * Usage:
 *   const manager = new ChallengeManager(gameCode, currentPlayer, gameAPI, gameWS);
 *   await manager.initialize();
 *   manager.onChallengesUpdated((challenges) => { updateUI(challenges); });
 */

class ChallengeManager {
    constructor(gameCode, currentPlayer, gameAPI, gameWS) {
        this.gameCode = gameCode;
        this.currentPlayer = currentPlayer;
        this.gameAPI = gameAPI;
        this.gameWS = gameWS;
        
        // State
        this.challenges = new Map(); // Map<challengeKey, challengeData>
        this.gameStatus = 'waiting'; // 'waiting', 'in_progress', 'paused', 'completed'
        this.lastPauseTime = null;
        
        // Callbacks
        this.updateCallbacks = [];
        this.timerInterval = null;
        
        // Constants
        this.CHALLENGE_DURATION_MS = 10 * 60 * 1000; // 10 minutes
        this.EXPIRY_CHECK_INTERVAL_MS = 1000; // Check every second
        
        // Bind WebSocket handler
        this._setupWebSocketHandlers();
    }
    
    /**
     * Initialize the challenge manager by loading challenges from server
     */
    async initialize() {
        // console.log('[ChallengeManager] Initializing...');
        await this.loadFromServer();
        this.startTimerInterval();
        // console.log('[ChallengeManager] Initialized with', this.challenges.size, 'challenges');
    }
    
    /**
     * Load all active challenges from server
     */
    async loadFromServer() {
        try {
            const challenges = await this.gameAPI.getChallenges(this.gameCode);
            // console.log('[ChallengeManager] loadFromServer - Raw challenges from API:', challenges);
            // console.log('[ChallengeManager] loadFromServer - Number of challenges:', challenges.length);
            
            // Fetch player names to enrich challenge data
            const players = await this.gameAPI.getPlayers(this.gameCode);
            // console.log('[ChallengeManager] loadFromServer - Players from API:', players);
            // console.log('[ChallengeManager] loadFromServer - First player structure:', players[0]);
            const playerMap = new Map(players.map(p => [p.id, p.name || p.player_name]));
            
            // Clear and rebuild challenge map
            this.challenges.clear();
            
            for (const challenge of challenges) {
                // console.log('[ChallengeManager] loadFromServer - Processing challenge:', challenge);
                // console.log('[ChallengeManager] loadFromServer - Challenge ID:', challenge.id);
                // console.log('[ChallengeManager] loadFromServer - Challenge status:', challenge.status);
                // console.log('[ChallengeManager] loadFromServer - Challenge player_id:', challenge.player_id);
                // console.log('[ChallengeManager] loadFromServer - Challenge team_number:', challenge.team_number);
                // console.log('[ChallengeManager] loadFromServer - Challenge building_type:', challenge.building_type);
                // console.log('[ChallengeManager] loadFromServer - Challenge assigned_at:', challenge.assigned_at);
                // console.log('[ChallengeManager] loadFromServer - Challenge start_time:', challenge.start_time);
                
                // Only load active challenges
                if (challenge.status === 'requested' || challenge.status === 'assigned') {
                    const key = this._getChallengeKey(challenge);
                    // console.log('[ChallengeManager] loadFromServer - Challenge key:', key);
                    const normalizedChallenge = this._normalizeChallenge(challenge);
                    // console.log('[ChallengeManager] loadFromServer - Normalized challenge:', normalizedChallenge);
                    // console.log('[ChallengeManager] loadFromServer - Normalized status:', normalizedChallenge.status);
                    
                    // Add player name from players list
                    normalizedChallenge.player_name = playerMap.get(challenge.player_id) || 'Unknown Player';
                    this.challenges.set(key, normalizedChallenge);
                }
            }
            
            this._notifyUpdates();
            // console.log('[ChallengeManager] Loaded', this.challenges.size, 'challenges from server');
            // console.log('[ChallengeManager] Final challenges map:', Array.from(this.challenges.entries()));
        } catch (error) {
            console.error('[ChallengeManager] Failed to load challenges:', error);
            throw error;
        }
    }
    
    /**
     * Request a new challenge (player action)
     */
    async requestChallenge(playerId, buildingType, buildingName, teamNumber, hasSchool) {
        try {
            // console.log('[ChallengeManager] Requesting challenge:', { playerId, buildingType });
            
            const response = await fetch(`${this.gameAPI.baseUrl}/games/${this.gameCode}/challenges`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.gameAPI.token}`
                },
                body: JSON.stringify({
                    player_id: playerId,
                    building_type: buildingType,
                    building_name: buildingName,
                    team_number: teamNumber,
                    has_school: hasSchool
                })
            });
            
            if (!response.ok) {
                throw new Error(`Failed to request challenge: ${response.statusText}`);
            }
            
            const challenge = await response.json();
            
            // Add to local state (WebSocket will sync with others)
            const key = this._getChallengeKey(challenge);
            this.challenges.set(key, this._normalizeChallenge(challenge));
            this._notifyUpdates();
            
            return challenge;
        } catch (error) {
            console.error('[ChallengeManager] Request failed:', error);
            throw error;
        }
    }
    
    /**
     * Assign a challenge (host/banker action)
     */
    async assignChallenge(challengeId, challengeType, challengeDescription, targetNumber) {
        try {
            // console.log('[ChallengeManager] Assigning challenge:', challengeId);
            
            const response = await this.gameAPI.updateChallenge(this.gameCode, challengeId, {
                status: 'assigned',
                challenge_type: challengeType,
                challenge_description: challengeDescription,
                target_number: targetNumber
            });
            
            // Update local state
            const key = this._findChallengeKeyById(challengeId);
            if (key) {
                const challenge = this.challenges.get(key);
                challenge.status = 'assigned';
                challenge.challenge_type = challengeType;
                challenge.challenge_description = challengeDescription;
                challenge.target_number = targetNumber;
                challenge.start_time = Date.now();
                challenge.assigned_at = new Date().toISOString();
                this.challenges.set(key, challenge);
                this._notifyUpdates();
            }
            
            return response;
        } catch (error) {
            console.error('[ChallengeManager] Assignment failed:', error);
            throw error;
        }
    }
    
    /**
     * Complete a challenge (host/banker action)
     */
    async completeChallenge(challengeId) {
        try {
            // console.log('[ChallengeManager] Completing challenge:', challengeId);
            
            await this.gameAPI.updateChallenge(this.gameCode, challengeId, {
                status: 'completed'
            });
            
            // Remove from local state
            const key = this._findChallengeKeyById(challengeId);
            if (key) {
                this.challenges.delete(key);
                this._notifyUpdates();
            }
            
            return { success: true };
        } catch (error) {
            console.error('[ChallengeManager] Completion failed:', error);
            throw error;
        }
    }
    
    /**
     * Cancel a challenge (host/banker action)
     */
    async cancelChallenge(challengeId) {
        try {
            // console.log('[ChallengeManager] Cancelling challenge:', challengeId);
            
            await this.gameAPI.updateChallenge(this.gameCode, challengeId, {
                status: 'cancelled'
            });
            
            // Remove from local state
            const key = this._findChallengeKeyById(challengeId);
            if (key) {
                this.challenges.delete(key);
                this._notifyUpdates();
            }
            
            return { success: true };
        } catch (error) {
            console.error('[ChallengeManager] Cancellation failed:', error);
            throw error;
        }
    }
    
    /**
     * Adjust challenge times for pause (called on resume)
     */
    async adjustForPause(pauseDurationMs) {
        try {
            // console.log('[ChallengeManager] Adjusting for pause:', pauseDurationMs, 'ms');
            
            const response = await fetch(`${this.gameAPI.baseUrl}/games/${this.gameCode}/challenges/adjust-for-pause`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.gameAPI.token}`
                },
                body: JSON.stringify({ pause_duration_ms: pauseDurationMs })
            });
            
            if (!response.ok) {
                throw new Error(`Failed to adjust challenges: ${response.statusText}`);
            }
            
            // Wait a moment for DB to commit
            await new Promise(resolve => setTimeout(resolve, 100));
            
            // Reload from server to get updated timestamps
            await this.loadFromServer();
            
            return await response.json();
        } catch (error) {
            console.error('[ChallengeManager] Pause adjustment failed:', error);
            throw error;
        }
    }
    
    /**
     * Get challenges for display (filtered by user role)
     */
    getChallengesForUser() {
        // console.log('[ChallengeManager] getChallengesForUser - currentPlayer:', this.currentPlayer);
        const isHostOrBanker = this.currentPlayer.role === 'host' || this.currentPlayer.role === 'banker';
        const challenges = Array.from(this.challenges.values());
        // console.log('[ChallengeManager] getChallengesForUser - all challenges:', challenges);
        // console.log('[ChallengeManager] getChallengesForUser - isHostOrBanker:', isHostOrBanker);
        
        if (isHostOrBanker) {
            // Host/banker sees all challenges
            // console.log('[ChallengeManager] getChallengesForUser - returning all challenges (host/banker)');
            return challenges;
        } else {
            // Team members see only their team's challenges
            // console.log('[ChallengeManager] getChallengesForUser - filtering by team:', this.currentPlayer.groupNumber);
            const filtered = challenges.filter(c => {
                // console.log('[ChallengeManager] getChallengesForUser - checking challenge team:', c.team_number, 'vs player team:', this.currentPlayer.groupNumber);
                return c.team_number === this.currentPlayer.groupNumber;
            });
            // console.log('[ChallengeManager] getChallengesForUser - filtered challenges:', filtered);
            return filtered;
        }
    }
    
    /**
     * Get only assigned challenges (for Active Challenges tab)
     */
    getAssignedChallenges() {
        const userChallenges = this.getChallengesForUser();
        // console.log('[ChallengeManager] getAssignedChallenges - user challenges:', userChallenges);
        // console.log('[ChallengeManager] getAssignedChallenges - current player role:', this.currentPlayer.role);
        // console.log('[ChallengeManager] getAssignedChallenges - current player groupNumber:', this.currentPlayer.groupNumber);
        
        const assigned = userChallenges.filter(c => {
            // console.log('[ChallengeManager] Checking challenge:', c);
            // console.log('[ChallengeManager] - status:', c.status, 'expected: assigned');
            // console.log('[ChallengeManager] - start_time:', c.start_time);
            return c.status === 'assigned' && c.start_time;
        });
        
        // console.log('[ChallengeManager] getAssignedChallenges - filtered assigned challenges:', assigned);
        return assigned;
    }
    
    /**
     * Get only requested challenges (for Challenge Requests tab)
     */
    getRequestedChallenges() {
        // console.log('[ChallengeManager] getRequestedChallenges called');
        // console.log('[ChallengeManager] currentPlayer:', this.currentPlayer);
        // console.log('[ChallengeManager] currentPlayer.role:', this.currentPlayer?.role);
        // console.log('[ChallengeManager] typeof role:', typeof this.currentPlayer?.role);
        
        // Normalize role to lowercase for comparison
        const role = this.currentPlayer?.role?.toLowerCase();
        const isHostOrBanker = role === 'host' || role === 'banker';
        // console.log('[ChallengeManager] normalized role:', role);
        // console.log('[ChallengeManager] isHostOrBanker:', isHostOrBanker);
        
        if (!isHostOrBanker) {
            // console.log('[ChallengeManager] Not host or banker, returning empty array');
            return [];
        }
        
        const requested = Array.from(this.challenges.values()).filter(c => c.status === 'requested');
        // console.log('[ChallengeManager] Requested challenges:', requested);
        // console.log('[ChallengeManager] Challenges map size:', this.challenges.size);
        // console.log('[ChallengeManager] All challenges:', Array.from(this.challenges.entries()));
        return requested;
    }
    
    /**
     * Calculate time remaining for a challenge
     */
    getTimeRemaining(challenge) {
        if (!challenge.start_time) {
            return null;
        }
        
        let now = Date.now();
        
        // If game is paused, freeze time at pause moment
        if (this.gameStatus === 'paused' && this.lastPauseTime) {
            now = this.lastPauseTime;
        }
        
        const elapsed = now - challenge.start_time;
        const remaining = this.CHALLENGE_DURATION_MS - elapsed;
        
        return Math.max(0, remaining);
    }
    
    /**
     * Set game status (waiting, in_progress, paused, completed)
     */
    setGameStatus(status, pauseTime = null) {
        // console.log('[ChallengeManager] Game status:', status);
        this.gameStatus = status;
        
        if (status === 'paused') {
            this.lastPauseTime = pauseTime || Date.now();
        } else if (status === 'in_progress') {
            this.lastPauseTime = null;
        }
    }
    
    /**
     * Register a callback for when challenges are updated
     */
    onChallengesUpdated(callback) {
        this.updateCallbacks.push(callback);
    }
    
    /**
     * Clean up resources
     */
    async clearAll() {
        // console.log('[ChallengeManager] Clearing all challenges');
        this.challenges.clear();
        this._notifyUpdates();
    }
    
    destroy() {
        if (this.timerInterval) {
            clearInterval(this.timerInterval);
            this.timerInterval = null;
        }
        this.updateCallbacks = [];
        this.challenges.clear();
    }
    
    // Private methods
    
    _getChallengeKey(challenge) {
        if (challenge.has_school) {
            return `${challenge.player_id}-${challenge.building_type}`;
        } else {
            return `team${challenge.team_number}-${challenge.building_type}`;
        }
    }
    
    _findChallengeKeyById(challengeId) {
        for (const [key, challenge] of this.challenges.entries()) {
            if (challenge.db_id === challengeId) {
                return key;
            }
        }
        return null;
    }
    
    _normalizeChallenge(challenge) {
        return {
            db_id: challenge.id,
            player_id: challenge.player_id,
            team_number: challenge.team_number,
            building_type: challenge.building_type,
            building_name: challenge.building_name,
            has_school: challenge.has_school,
            status: challenge.status,
            challenge_type: challenge.challenge_type || null,
            challenge_description: challenge.challenge_description || null,
            target_number: challenge.target_number || null,
            start_time: challenge.assigned_at ? new Date(challenge.assigned_at).getTime() : null,
            timestamp: challenge.requested_at || challenge.assigned_at || new Date().toISOString(),
            requested_at: challenge.requested_at,
            assigned_at: challenge.assigned_at,
            completed_at: challenge.completed_at
        };
    }
    
    _notifyUpdates() {
        // console.log('[ChallengeManager] _notifyUpdates called, callbacks count:', this.updateCallbacks.length);
        // console.log('[ChallengeManager] Current challenges:', Array.from(this.challenges.entries()));
        for (const callback of this.updateCallbacks) {
            try {
                callback(this.challenges);
            } catch (error) {
                console.error('[ChallengeManager] Update callback error:', error);
            }
        }
    }
    
    _setupWebSocketHandlers() {
        // Hook into existing WebSocket event system
        // This assumes the WebSocket has an event handler we can extend
        // console.log('[ChallengeManager] Setting up WebSocket handlers');
    }
    
    startTimerInterval() {
        // Clear existing interval
        if (this.timerInterval) {
            clearInterval(this.timerInterval);
        }
        
        // Start new interval
        this.timerInterval = setInterval(() => {
            this._checkExpiry();
            // Also notify updates every second to refresh countdown timers in UI
            // Only if there are active challenges and game is in progress
            if (this.gameStatus === 'in_progress' && this.getAssignedChallenges().length > 0) {
                this._notifyUpdates();
            }
        }, this.EXPIRY_CHECK_INTERVAL_MS);
    }
    
    _checkExpiry() {
        // Only expire challenges when game is actively running
        if (this.gameStatus !== 'in_progress') {
            return;
        }
        
        const now = Date.now();
        let expiredAny = false;
        
        for (const [key, challenge] of this.challenges.entries()) {
            if (challenge.status === 'assigned' && challenge.start_time) {
                const elapsed = now - challenge.start_time;
                
                if (elapsed >= this.CHALLENGE_DURATION_MS) {
                    // console.log('[ChallengeManager] Challenge expired:', key);
                    
                    // Mark as expired in database
                    this.gameAPI.updateChallenge(this.gameCode, challenge.db_id, { status: 'expired' })
                        .catch(err => console.error('[ChallengeManager] Failed to expire challenge:', err));
                    
                    // Remove from local state
                    this.challenges.delete(key);
                    expiredAny = true;
                }
            }
        }
        
        // No need to call _notifyUpdates here since the interval already does it every second
        // This avoids double notifications
    }
    
    // WebSocket event handlers (to be called from main dashboard code)
    
    handleChallengeRequest(eventData) {
        // console.log('[ChallengeManager] WebSocket: challenge_request', eventData);
        // console.log('[ChallengeManager] Current player:', this.currentPlayer);
        // console.log('[ChallengeManager] Current player role:', this.currentPlayer?.role);
        
        const challenge = {
            db_id: eventData.db_id || null, // Database ID from HTTP response
            player_id: eventData.player_id,
            player_name: eventData.player_name,
            team_number: eventData.team_number,
            building_type: eventData.building_type,
            building_name: eventData.building_name,
            has_school: eventData.has_school,
            status: 'requested',
            challenge_type: null,
            challenge_description: null,
            target_number: null,
            start_time: null,
            timestamp: new Date().toISOString(),
            requested_at: new Date().toISOString(),
            assigned_at: null,
            completed_at: null
        };
        
        const key = this._getChallengeKey(challenge);
        // console.log('[ChallengeManager] Adding challenge with key:', key);
        // console.log('[ChallengeManager] Challenge object:', challenge);
        this.challenges.set(key, challenge);
        // console.log('[ChallengeManager] Challenges map after add:', Array.from(this.challenges.entries()));
        // console.log('[ChallengeManager] Challenges map size:', this.challenges.size);
        // console.log('[ChallengeManager] Calling _notifyUpdates()...');
        this._notifyUpdates();
        // console.log('[ChallengeManager] _notifyUpdates() called');
    }
    
    handleChallengeAssigned(eventData) {
        // console.log('[ChallengeManager] WebSocket: challenge_assigned', eventData);
        
        // Find existing challenge
        const key = this._findChallengeKeyByPlayerAndBuilding(eventData.player_id, eventData.building_type);
        
        if (key) {
            const challenge = this.challenges.get(key);
            challenge.status = 'assigned';
            challenge.challenge_type = eventData.challenge_type;
            challenge.challenge_description = eventData.challenge_description;
            challenge.target_number = eventData.target_number;
            challenge.start_time = eventData.start_time;
            // Handle date conversion safely
            if (eventData.start_time) {
                const date = new Date(eventData.start_time);
                challenge.assigned_at = isNaN(date.getTime()) ? new Date().toISOString() : date.toISOString();
            } else {
                challenge.assigned_at = new Date().toISOString();
            }
            this.challenges.set(key, challenge);
            this._notifyUpdates();
        } else {
            // Challenge doesn't exist locally - reload from server
            // console.log('[ChallengeManager] Challenge not found locally, reloading from server');
            this.loadFromServer();
        }
    }
    
    handleChallengeCompleted(eventData) {
        // console.log('[ChallengeManager] WebSocket: challenge_completed', eventData);
        
        const key = this._findChallengeKeyByPlayerAndBuilding(eventData.player_id, eventData.building_type);
        if (key) {
            this.challenges.delete(key);
            this._notifyUpdates();
        }
    }
    
    handleChallengeCancelled(eventData) {
        // console.log('[ChallengeManager] WebSocket: challenge_cancelled', eventData);
        
        const key = this._findChallengeKeyByPlayerAndBuilding(eventData.player_id, eventData.building_type);
        if (key) {
            this.challenges.delete(key);
            this._notifyUpdates();
        }
    }
    
    _findChallengeKeyByPlayerAndBuilding(playerId, buildingType) {
        for (const [key, challenge] of this.challenges.entries()) {
            if (challenge.player_id === playerId && challenge.building_type === buildingType) {
                return key;
            }
        }
        return null;
    }
}

// Export for use in dashboard
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ChallengeManager;
}

File: backend/challenge_manager.py
"""
Challenge Management Service

Handles all challenge lifecycle operations with pause-aware timing.
"""
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from models import Challenge, ChallengeStatus, GameSession, Player
from websocket_manager import manager as ws_manager


class ChallengeManager:
    """
    Centralized challenge management with pause-aware timing.
    
    Key features:
    - Single source of truth for challenge state
    - Automatic expiry checking with pause compensation
    - Multi-user (host/banker) support
    - Database-backed state management
    """
    
    CHALLENGE_DURATION_MINUTES = 10
    
    def __init__(self, db: Session):
        self.db = db
    
    async def create_challenge_request(
        self,
        game_code: str,
        player_id: int,
        building_type: str,
        building_name: str,
        team_number: int,
        has_school: bool
    ) -> Challenge:
        """
        Create a new challenge request from a player.
        
        Args:
            game_code: Game session code
            player_id: ID of requesting player
            building_type: Type of building (farm, mine, etc.)
            building_name: Formatted building name
            team_number: Player's team number
            has_school: Whether team has a school (affects locking)
        
        Returns:
            Created Challenge object
        """
        game = self.db.query(GameSession).filter(
            GameSession.game_code == game_code.upper()
        ).first()
        
        if not game:
            raise ValueError(f"Game {game_code} not found")
        
        # Check if there's already an active challenge for this player/building
        existing = self.db.query(Challenge).filter(
            Challenge.game_session_id == game.id,
            Challenge.player_id == player_id,
            Challenge.building_type == building_type,
            Challenge.status.in_([ChallengeStatus.REQUESTED, ChallengeStatus.ASSIGNED])
        ).first()
        
        if existing:
            raise ValueError(f"Challenge already exists for player {player_id} building {building_type}")
        
        # Check bank inventory for required resources
        team_key = str(team_number)
        team_data = game.game_state.get('teams', {}).get(team_key, {}) if game.game_state else {}
        building_count = team_data.get('buildings', {}).get(building_type, 0)
        
        # Map building types to resources
        production_grants = {
            'farm': {'resource': 'food', 'amount': 5},
            'mine': {'resource': 'raw_materials', 'amount': 5},
            'electrical_factory': {'resource': 'electrical_goods', 'amount': 5},
            'medical_factory': {'resource': 'medical_goods', 'amount': 5}
        }
        
        grant_info = production_grants.get(building_type)
        if not grant_info:
            raise ValueError(f"Invalid building type: {building_type}")
        
        required_resource = grant_info['resource']
        base_amount = grant_info['amount']
        required_amount = base_amount * building_count * 1.0  # Normal difficulty
        
        # Check bank inventory from game_state
        if game.game_state and 'bank_inventory' in game.game_state:
            bank_inventory = game.game_state.get('bank_inventory', {})
            current_inventory = bank_inventory.get(required_resource, 0)
            
            if current_inventory < required_amount:
                raise ValueError(
                    f"Bank does not have enough {required_resource}. "
                    f"Required: {int(required_amount)}, Available: {int(current_inventory)}"
                )
        
        challenge = Challenge(
            game_session_id=game.id,
            player_id=player_id,
            building_type=building_type,
            building_name=building_name,
            team_number=team_number,
            has_school=has_school,
            status=ChallengeStatus.REQUESTED,
            requested_at=datetime.utcnow()
        )
        
        self.db.add(challenge)
        self.db.commit()
        self.db.refresh(challenge)
        
        # Broadcast to all connected clients (especially host/banker)
        await ws_manager.broadcast_challenge_requested(
            game_code.upper(),
            self.serialize_challenge(challenge)
        )
        
        return challenge
    
    async def assign_challenge(
        self,
        challenge_id: int,
        challenge_type: str,
        challenge_description: str,
        target_number: int
    ) -> Challenge:
        """
        Assign a challenge to a player (host/banker action).
        
        Args:
            challenge_id: ID of the challenge to assign
            challenge_type: Type of physical challenge (push_ups, etc.)
            challenge_description: Full description ("20 Push-ups")
            target_number: Target count
        
        Returns:
            Updated Challenge object with assigned_at timestamp
        """
        challenge = self.db.query(Challenge).filter(
            Challenge.id == challenge_id
        ).first()
        
        if not challenge:
            raise ValueError(f"Challenge {challenge_id} not found")
        
        if challenge.status != ChallengeStatus.REQUESTED:
            raise ValueError(f"Challenge {challenge_id} is not in REQUESTED state")
        
        challenge.status = ChallengeStatus.ASSIGNED
        challenge.challenge_type = challenge_type
        challenge.challenge_description = challenge_description
        challenge.target_number = target_number
        challenge.assigned_at = datetime.utcnow()
        
        self.db.commit()
        self.db.refresh(challenge)
        
        # Broadcast to all connected clients
        game = challenge.game_session
        await ws_manager.broadcast_challenge_assigned(
            game_code=game.game_code,
            challenge_data=self.serialize_challenge(challenge, include_time_remaining=True)
        )
        
        # Send notification to the team
        player = challenge.player
        if player and player.group_number:
            challenge_type_name = challenge_type.replace('_', ' ').title()
            notification_message = f"Challenge assigned: {challenge_description} for {challenge.building_type.replace('_', ' ').title()}"
            
            await ws_manager.send_to_team(
                game_code=game.game_code,
                team_number=player.group_number,
                message={
                    "type": "notification",
                    "message": notification_message,
                    "notification_type": "challenge_assigned",
                    "challenge_data": self.serialize_challenge(challenge, include_time_remaining=True)
                },
                db_session=self.db
            )
        
        return challenge
    
    async def complete_challenge(self, challenge_id: int) -> Challenge:
        """
        Mark a challenge as completed.
        
        Args:
            challenge_id: ID of the challenge to complete
        
        Returns:
            Updated Challenge object
        """
        challenge = self.db.query(Challenge).filter(
            Challenge.id == challenge_id
        ).first()
        
        if not challenge:
            raise ValueError(f"Challenge {challenge_id} not found")
        
        if challenge.status != ChallengeStatus.ASSIGNED:
            raise ValueError(f"Challenge {challenge_id} is not in ASSIGNED state")
        
        challenge.status = ChallengeStatus.COMPLETED
        challenge.completed_at = datetime.utcnow()
        
        self.db.commit()
        self.db.refresh(challenge)
        
        # Broadcast completion to all connected clients
        game = challenge.game_session
        await ws_manager.broadcast_challenge_completed(
            game_code=game.game_code,
            challenge_data=self.serialize_challenge(challenge)
        )
        
        # Send notification to the team
        player = challenge.player
        if player and player.group_number:
            building_name = challenge.building_type.replace('_', ' ').title()
            
            # Determine the resource and amount earned
            production_grants = {
                'farm': {'resource': 'food', 'amount': 5},
                'mine': {'resource': 'raw_materials', 'amount': 5},
                'electrical_factory': {'resource': 'electrical_goods', 'amount': 5},
                'medical_factory': {'resource': 'medical_goods', 'amount': 5}
            }
            
            grant_info = production_grants.get(challenge.building_type, {'resource': 'resources', 'amount': 5})
            resource_name = grant_info['resource'].replace('_', ' ').title()
            resource_amount = grant_info['amount']
            
            notification_message = f"✅ {player.player_name} completed {challenge.challenge_description} and earned {resource_amount} {resource_name}!"
            
            await ws_manager.send_to_team(
                game_code=game.game_code,
                team_number=player.group_number,
                message={
                    "type": "notification",
                    "message": notification_message,
                    "notification_type": "challenge_completed",
                    "challenge_data": self.serialize_challenge(challenge)
                },
                db_session=self.db
            )
        
        return challenge
    
    async def cancel_challenge(self, challenge_id: int) -> Challenge:
        """
        Cancel a challenge (can be requested or assigned).
        
        Args:
            challenge_id: ID of the challenge to cancel
        
        Returns:
            Updated Challenge object
        """
        challenge = self.db.query(Challenge).filter(
            Challenge.id == challenge_id
        ).first()
        
        if not challenge:
            raise ValueError(f"Challenge {challenge_id} not found")
        
        if challenge.status in [ChallengeStatus.COMPLETED, ChallengeStatus.CANCELLED]:
            raise ValueError(f"Challenge {challenge_id} is already {challenge.status.value}")
        
        challenge.status = ChallengeStatus.CANCELLED
        
        self.db.commit()
        self.db.refresh(challenge)
        
        # Broadcast cancellation to all connected clients
        game = challenge.game_session
        await ws_manager.broadcast_challenge_cancelled(
            game_code=game.game_code,
            challenge_data=self.serialize_challenge(challenge)
        )
        
        return challenge
    
    def adjust_for_pause(self, game_code: str, pause_duration_ms: int) -> Dict[str, Any]:
        """
        Adjust all active challenge timestamps to account for pause duration.
        This extends the deadline by adding the pause duration to assigned_at.
        
        Args:
            game_code: Game session code
            pause_duration_ms: Duration of pause in milliseconds
        
        Returns:
            Dict with adjustment details
        """
        game = self.db.query(GameSession).filter(
            GameSession.game_code == game_code.upper()
        ).first()
        
        if not game:
            raise ValueError(f"Game {game_code} not found")
        
        # Get all assigned challenges
        active_challenges = self.db.query(Challenge).filter(
            Challenge.game_session_id == game.id,
            Challenge.status == ChallengeStatus.ASSIGNED,
            Challenge.assigned_at.isnot(None)
        ).all()
        
        if not active_challenges:
            return {
                "success": True,
                "adjusted_count": 0,
                "pause_duration_ms": pause_duration_ms
            }
        
        pause_duration = timedelta(milliseconds=pause_duration_ms)
        adjusted_count = 0
        
        for challenge in active_challenges:
            original_time = challenge.assigned_at
            challenge.assigned_at = challenge.assigned_at + pause_duration
            adjusted_count += 1
        
        self.db.commit()
        
        return {
            "success": True,
            "adjusted_count": adjusted_count,
            "pause_duration_ms": pause_duration_ms
        }
    
    def check_and_expire_challenges(self, game_code: str) -> List[Challenge]:
        """
        Check all assigned challenges and expire those past their deadline.
        
        Args:
            game_code: Game session code
        
        Returns:
            List of expired challenges
        """
        game = self.db.query(GameSession).filter(
            GameSession.game_code == game_code.upper()
        ).first()
        
        if not game:
            raise ValueError(f"Game {game_code} not found")
        
        # Get all assigned challenges
        active_challenges = self.db.query(Challenge).filter(
            Challenge.game_session_id == game.id,
            Challenge.status == ChallengeStatus.ASSIGNED,
            Challenge.assigned_at.isnot(None)
        ).all()
        
        expired_challenges = []
        now = datetime.utcnow()
        expiry_duration = timedelta(minutes=self.CHALLENGE_DURATION_MINUTES)
        
        for challenge in active_challenges:
            deadline = challenge.assigned_at + expiry_duration
            if now >= deadline:
                challenge.status = ChallengeStatus.EXPIRED
                expired_challenges.append(challenge)
        
        if expired_challenges:
            self.db.commit()
        
        return expired_challenges
    
    def get_active_challenges(self, game_code: str) -> List[Challenge]:
        """
        Get all active (requested or assigned) challenges for a game.
        
        Args:
            game_code: Game session code
        
        Returns:
            List of active challenges
        """
        game = self.db.query(GameSession).filter(
            GameSession.game_code == game_code.upper()
        ).first()
        
        if not game:
            raise ValueError(f"Game {game_code} not found")
        
        challenges = self.db.query(Challenge).filter(
            Challenge.game_session_id == game.id,
            Challenge.status.in_([ChallengeStatus.REQUESTED, ChallengeStatus.ASSIGNED])
        ).order_by(Challenge.requested_at.desc()).all()
        
        return challenges
    
    def get_challenge_time_remaining(self, challenge: Challenge) -> Optional[int]:
        """
        Calculate remaining time for a challenge in seconds.
        
        Args:
            challenge: Challenge object
        
        Returns:
            Remaining seconds, or None if not assigned
        """
        if challenge.status != ChallengeStatus.ASSIGNED or not challenge.assigned_at:
            return None
        
        now = datetime.utcnow()
        expiry_duration = timedelta(minutes=self.CHALLENGE_DURATION_MINUTES)
        deadline = challenge.assigned_at + expiry_duration
        remaining = deadline - now
        
        return max(0, int(remaining.total_seconds()))
    
    def serialize_challenge(self, challenge: Challenge, include_time_remaining: bool = False) -> Dict[str, Any]:
        """
        Serialize a challenge to a dict for API responses.
        
        Args:
            challenge: Challenge object
            include_time_remaining: Whether to include calculated time remaining
        
        Returns:
            Dict representation of challenge
        """
        # Get player name from relationship
        player_name = challenge.player.player_name if challenge.player else "Unknown"
        
        data = {
            "id": challenge.id,
            "game_session_id": challenge.game_session_id,
            "player_id": challenge.player_id,
            "player_name": player_name,
            "building_type": challenge.building_type,
            "building_name": challenge.building_name,
            "team_number": challenge.team_number,
            "has_school": challenge.has_school,
            "challenge_type": challenge.challenge_type,
            "challenge_description": challenge.challenge_description,
            "target_number": challenge.target_number,
            "status": challenge.status.value,
            "requested_at": challenge.requested_at.isoformat() if challenge.requested_at else None,
            "assigned_at": challenge.assigned_at.isoformat() if challenge.assigned_at else None,
            "completed_at": challenge.completed_at.isoformat() if challenge.completed_at else None,
            "timestamp": challenge.requested_at.isoformat() if challenge.requested_at else None,  # For frontend compatibility
            "start_time": int(challenge.assigned_at.timestamp() * 1000) if challenge.assigned_at else None,  # Milliseconds for JS
        }
        
        if include_time_remaining:
            data["time_remaining_seconds"] = self.get_challenge_time_remaining(challenge)
        
        return data
